// @ifi/oh-pi-core/dist/i18n.js
import { execSync } from "node:child_process";
import * as p from "@clack/prompts";
import { messages } from "./locales.js";
let current = "en";
/**
 * Get a translated string for the current locale, with optional variable interpolation.
 */
export function t(key, vars) {
    let text = messages[current]?.[key] ?? messages.en[key] ?? key;
    if (vars) {
        for (const [k, v] of Object.entries(vars)) {
            text = text.replace(`{${k}}`, String(v));
        }
    }
    return text;
}
/** Set the current locale. */
export function setLocale(locale) {
    current = locale;
}
/** Get the current locale. */
export function getLocale() {
    return current;
}
/** Detect the user's locale from environment variables. */
function detectLocale() {
    let lang = (process.env.LANG ?? process.env.LC_ALL ?? process.env.LANGUAGE ?? "").toLowerCase();
    if (!lang && process.platform === "win32") {
        try {
            lang = execSync('powershell -NoProfile -Command "(Get-Culture).Name"', { encoding: "utf8", timeout: 3000 })
                .trim()
                .toLowerCase();
        }
        catch {
            /* ignore */
        }
    }
    if (lang.startsWith("fr")) {
        return "fr";
    }
    if (lang.startsWith("en")) {
        return "en";
    }
    return undefined;
}
/** Prompt the user to select a language. Auto-detects from environment if possible, otherwise shows an interactive selector. */
export async function selectLanguage() {
    const detected = detectLocale();
    if (detected) {
        setLocale(detected);
        return detected;
    }
    const locale = await p.select({
        message: "Language / Langue:",
        options: [
            { value: "en", label: "English" },
            { value: "fr", label: "Français" },
        ],
    });
    if (p.isCancel(locale)) {
        p.cancel("Cancelled.");
        process.exit(0);
    }
    setLocale(locale);
    return locale;
}
